### PR TITLE
Types for gulp-tap

### PR DESCRIPTION
**Pull request by TokugawaTakeshi**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`

Warning:

Because I don't know, what means optional parameter `t` in below code from example in readme of [gulp-tap](https://www.npmjs.com/package/gulp-tap) package, I added this parameter as `t` with `any` type. 

```javascript
gulp.src("src/**/*.{coffee,js}")
    .pipe(tap(function(file, t) {
        if (path.extname(file.path) === '.coffee') {
            return t.through(coffee, []);
        }
    }))
    .pipe(gulp.dest('build'));
```

Unfortunately, author of gulp-tap is not answering for now. If he will answer with explanations about `t` parameter, I'll edit my definitions and create another pull request.

### CODE REVIEW

File: types/gulp-tap/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    interface Tap {
+        (tapFunction: TapFunction, t?: any): NodeJS.ReadWriteStream;
+    }
+
+    interface TapFunction {
+        (file: Vinyl): any;
+    }
+}
+
+declare var tap: tap.Tap;
```
Comment (sheetalkamat) [inline types/gulp-tap/index.d.ts]: This probably needs to be declared as:
```ts
declare function tap(tapFn: (file: Vinyl, t?: {}) => void): NodeJS.ReadWriteStream;
```

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

**Comment by TokugawaTakesi** [comment]
I am clear. Please, check.

**Comment by TokugawaTakesi** [comment]
@sheetalkamat, would you please to review my changes?

File: types/gulp-tap/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,23 @@
+{
+    "compilerOptions": {
+        "module": "commonjs",
+        "lib": [
+            "es6"
+        ],
+        "noImplicitAny": true,
+        "noImplicitThis": false,
+        "strictNullChecks": false,
```
Comment (sheetalkamat) [inline types/gulp-tap/tsconfig.json]: This cannot be off in new definition.

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

**Comment by TokugawaTakesi** [comment]
@sheetalkamat, I am clear. Please, check.

**Review by sheetalkamat — APPROVED** [review]

**Pull request merged**